### PR TITLE
avoid repeat connections in pings every round

### LINKED ISSUE #21739: FileBasedDiscovery causes connections on every ping round
Currently `UnicastHostsProvider#buildDynamicNodes` is called multiple times during the pinging rounds. Once on start, one more after 1.5 seconds (by default) and last time at the end (3s by default).  My guess is that this done in an attempt to get the latest information possible when connect to EC2 / GCE/ Azure's API. This does have a nasty side effect on the FileBasedUnicastHostsProvider, which currently reads it's file on ever call. The file reading is not a problem on it's own as it is just as fast as an EC2 call. The problem part is that each read generates nodes with new node ids, meaning that UnicastZenPing will setup new connections with each of these rounds. 

I see two relatively easy ways of solving it:
1) Call `buildDynamicNodes` once at the beginning of each round (might be the simplest).
2) Add an "onStartPing" method to `UnicastHostsProvider` which is guaranteed to be called at the beginning of each pinging round. `FileBasedDiscovery` can read the file contents in that method.

### PR DESCRIPTION
**Pull request by chengpohi**
This PR is trying to fix: https://github.com/elastic/elasticsearch/issues/21739 by Option 1.

@bleskes Could you review this?

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java [hunk 1]
```diff
@@ -292,6 +292,7 @@ public void clearTemporalResponses() {
      */
     @Override
     public void ping(final PingListener listener, final TimeValue duration) {
+        final List<DiscoveryNode> sortedNodesToPing = buildNodesToPing();
```
Comment (bleskes) [inline core/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java]: can we unify the resolvedDiscoveryNodes logic with the buildNodesToPing ? They are similar in the sense that they don't change during a ping cycle.

**Review by bleskes — CHANGES_REQUESTED** [review]
Thx @chengpohi ! I left some comments. It would also be great to have a test that the unicast host provider is only called once per pinging round. Also it would be good to have a test that pinging picks up on nodes that it discovered - i.e., if you A & C ping B , A starts pinging C as well (it got it through B). See UnicastZenPingTests for examples and where this should go.

File: core/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java (hunk 1 above)
Comment (chengpohi) [inline core/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java]: Hi @bleskes, I have updated this base on your suggestion, Could you review again? Thanks

File: core/src/test/java/org/elasticsearch/discovery/zen/UnicastZenPingTests.java [hunk 2]
```diff
@@ -252,6 +254,151 @@ public ClusterState clusterState() {
...
+                return ClusterState.builder(state).blocks(ClusterBlocks.builder().addGlobalBlock(STATE_NOT_RECOVERED_BLOCK)).build();
+            }
+        });
+        closeables.push(zenPingA);
+
+        zenPingA.pingAndWait(TimeValue.timeValueSeconds(1));
+        verify(unicastHostsProviderA, times(1)).buildDynamicNodes();
+    }
+
+    public void testShouldPingDiscoveredNodes() throws IOException, InterruptedException {
```
Comment (bleskes) [inline core/src/test/java/org/elasticsearch/discovery/zen/UnicastZenPingTests.java]: can we call this `testDiscoveryOfNoneConfiguredNodes` and add a comment: `test that nodes discover each other if they ping a common host` ?

This doesn't test what I intended - i.e., test that a node pings nodes that have pinged it while it was pinging. That is however non trivial to test and I don't want to delay your change for setting this test up.

**Review by bleskes — CHANGES_REQUESTED** [review]
Thx @chengpohi . I left some more comments

File: core/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java [hunk 3]
```diff
@@ -518,6 +503,33 @@ public void run() {
...
+        nodesToPingSet.addAll(hostsProvider.buildDynamicNodes());
+
+        // add all possible master nodes that were active in the last known cluster configuration
+        for (ObjectCursor<DiscoveryNode> masterNode : contextProvider.nodes().getMasterNodes().values()) {
+            nodesToPingSet.add(masterNode.value);
+        }
+
+        final List<DiscoveryNode> resolvedDiscoveryNodes;
+        try {
+            resolvedDiscoveryNodes = resolveDiscoveryNodes(
```
Comment (chengpohi) [inline core/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java]: Hi @bleskes , if we add `resolveDiscoveryNodes`, we can't guarantee *configured hosts first* ping, Does it make sense?

**Comment by jasontedor** [comment]
Thanks for picking this up @chengpohi but this has now been superseded by #22277.

**Pull request closed without merging**